3.optimizing_CP/CellProfiler_optuna_utils/optuna_profiling_utils.py

```diff
@@ -0,0 +1,1157 @@
+
+        # replace Image_Metadata string in column names with Metadata_
+        annotated_df.columns = annotated_df.columns.str.replace(
+            "Image_Metadata_", "Metadata_"
+        )
+
+        # save annotated df as parquet file
+        output(
+            df=annotated_df,
+            output_filename=output_file_path,
+            output_type="parquet",
+        )
+        logger.info(f"Annotations have been added to {data_run} and saved!")
+        # check last annotated df to see if it has been annotated correctly
+        logger.info(annotated_df.shape)
+        annotated_df.head()
+    return output_file_path
+
+
+def retrieve_cell_count(
+    _path_df: pathlib.Path,
+    _read_specific_columns: bool = False,
+    _columns_to_read: list = None,
+    _groupby_columns: list = ["Metadata_Well"],
+    _columns_to_unique_count: list = [
+        "Metadata_Well",
+        "Metadata_Nuclei_TrackObjects_Label",
+    ],
+    _actual_column_to_sum: str = "Metadata_Nuclei_TrackObjects_Label",
+) -> tuple:
+    """This function retrieves the actual and target cell counts from a dataframe
+    The path supplied should be a parquet file that contains the cell counts from a CellProfiler pipeline
+    These actual and target cell counts are used later for the optimization using a loss function
+
+    Parameters
+    ----------
+    _path_df : pathlib.Path
+        This is the path to the parquet file that contains the cell counts
+    _read_specific_columns : bool, optional
+        This is an option to read the whole parquet file or only certain columns, by default False
+    _columns_to_read : list, optional
+        This is the specific columns in the parquet to read, by default None
+    _groupby_columns : list, optional
+        This is the specified columns to group by, by default ['Metadata_Well']
+    _columns_to_unique_count : list, optional
+        This is the columns to count the unique values of, by default [
+        "Metadata_Well",
+        "Metadata_Nuclei_TrackObjects_Label",
+    ]
+    _actual_column_to_sum : str, optional
+        This is the actual cell count column to calculate and sum not the target, by default "Metadata_Nuclei_TrackObjects_Label"
+
+    Returns
+    -------
+    tuple of lists
+        Returns a tuple of lists of the actual and target cell counts
+    """
+
+    # check if the columns to read are specified
+    if _read_specific_columns:
+        assert (
+            _columns_to_read is not None
+        ), "If read_specific_columns is True, columns_to_read must be specified"
+        # read in the dataframe
+        df = pd.read_parquet(_path_df, columns=_columns_to_read)
+    else:
+        # read in the dataframe
+        df = pd.read_parquet(_path_df)
+
+    logger.info(df.shape)
+
+    ########################################################
+    # Get the actual cell counts
+    ########################################################
```

Comment:
Super interesting-looking code comment! Do you find this to be more helpful?

Reply:
Yes :D